### ISSUE TITLE
Linters for HTML, JS

### ISSUE THREAD

**Issue by koeaw** [acdh-oeaw/apis-rdf-devops#21]
I've looked at the **precommit** config used for APIS vanilla but didn't see anything related to linting/code formatting for HTML or JavaScript in there.

Same as how we agreed on using [black](https://black.readthedocs.io/en/stable/) for Python, I think it would be useful if we all used consistent formatting for Django templates (and related files) also.

Whenever I've refactored templates now, for example, I've always reduced the indentation to 2 spaces first thing, because we currently seem to use old-school 4-space indentation everywhere, which very quickly makes HTML & JS files hard to parse/makes an IDE wrap lines a lot sooner. (2 spaces is best practice nowadays.)

**Comment by koeaw** [comment]
Django Packages suggests three tools for template linting: https://djangopackages.org/grids/g/template-linters/

**Comment by koeaw** [comment]
It may make sense to look at how other (larger) projects handle this issue.

From Wagtail's [HTML guidelines](https://docs.wagtail.org/en/stable/contributing/ui_guidelines.html):
> We use [djhtml](https://github.com/rtts/djhtml) for formatting and [Curlylint](https://www.curlylint.org/) for linting.

**Comment by koeaw** [comment]
Relatedly: acdh-oeaw/apis-core-rdf#42

I think a linter solution for templates – once we have one – should be applied to all relevant files in one go, just like the other issue proposes for Python code + black.

Manual updates result in overhead IME, because I'll often only think of reformatting a template for readability _after_ I've already started working on it for other reasons...

**Comment by steffres** [comment]
I agree. Please go ahead with this.

**Comment by koeaw** [comment]
Tagged for next team discussion in case someone else already has hands-on experience with this.

**Comment by sennierer** [comment]
@richardhadden to do some research and come up with suggestions for nex jfx

**Comment by koeaw** [comment]
Following on from my last comment in https://github.com/acdh-oeaw/apis-core-rdf/pull/116, which contains a GitHub workflow for reformatting templates using [djLint](https://github.com/Riverside-Healthcare/djLint) (+ reformatted templates):

> I ended up installing the package to work around the merge conflicts & to take a closer look at its capabilities in general because it would be preferable if it could be used with apis-rdf-devops / for local development

... Doing so, I noticed one thing it forces on users which is detrimental to readability, but for opting out of which there's already an open feature request, which is solvable (to the extent it's relevant for us) relatively easily, for which I'd want to submit a PR.

The only blocker for using DJLint with `apis-rdf-devops` seems to be the [pinned PyYAML version](https://github.com/acdh-oeaw/apis-core-rdf/blob/dev/pyproject.toml#L31) in `apis-core-rdf`. It caused Poetry to get stuck in a loop for me as I was trying to add djlint to `apis-rdf-devops`, which I worked around locally by manually changing the version no. in the `pyproject.toml` file. But I don't know if there's a way to test if/how this would affect other dependencies.

The GitHub workflow doesn't seem to use Poetry to install the package anyway, so it looks like we could use it independently, but I would suggest to use the same set of djLint settings across all repos. I'd also opt to put the settings in `.djlintrc` files to keep the  `pyproject.toml` clean / for greater transparency (probably even more applicable in the case of the GitHub workflow, which `pip` installs the package rather than using Poetry).

**Comment by b1rger** [comment]
> I noticed one thing it forces on users which is detrimental to readability, but for opting out of which there's already an open feature request, which is solvable 

and that would be?

> The only blocker for using DJLint with `apis-rdf-devops` seems to be the [pinned PyYAML version](https://github.com/acdh-oeaw/apis-core-rdf/blob/dev/pyproject.toml#L31)

I don't think it makes sense to implement this in `apis-rdf-devops`, given #59 - when that is done there won't be any template files left and this repository will be what its names says it is: a devops repo.

Also, the PyYAML version is not a blocker for using DJLint. Only if you want to use the same Python environment for both. But the PyYAML dependency is actually a very easy fix, given it is only used by `RDFParser` which is not working right now anyway, so we can simply update the dependency.

> The GitHub workflow doesn't seem to use Poetry to install the package anyway, so it looks like we could use it independently, but I would suggest to use the same set of djLint settings across all repos. I'd also opt to put the settings in `.djlintrc` files to keep the `pyproject.toml` clean / for greater transparency 

I am strongly against spreading configuration over multiple files if we have the chance to keep it in a single location. Most modern Python packaging tools are supporting `pyproject.toml` or are working on implementing support. Using one commonly used configuration file gives us the advantage of having one single source of truth. It also means we don't have to remember every configuration file location for every tool we ever want to use.
(Aside from `.djlintrc` being json...)

> (probably even more applicable in the case of the GitHub workflow, which `pip` installs the package rather than using Poetry).

I don't get what the way a software is installed has to do with where it gets its configuration settings from.

**Comment by koeaw** [comment]
> and that would be?

Its condensation "feature".

> I don't think it makes sense to implement this in apis-rdf-devops, given https://github.com/acdh-oeaw/apis-rdf-devops/issues/59 - when that is done there won't be any template files left and this repository will be what its names says it is: a devops repo.

Ok, but how would we use the tool for local development then? Analogous to black. Which is also part of `apis-rdf-devops`' dev dependencies, from where it can be used for all 3 repos. I'm assuming our own templates / template overrides will continue to reside in `apis-ontologies`, so what's the idea for how an HTML linter/formatter would be used so it reaches all relevant files across all repos/apps?

> I am strongly against spreading configuration over multiple files if we have the chance to keep it in a single location. 

Except we won't if we want to make use of djLint's rules, which need to go in a separate yml file. Which is similarly named to the config file, so those two files end up sitting right next to each other / are recognisable as belonging together.

> Most modern Python packaging tools are supporting `pyproject.toml` or are working on implementing support.

Ok, I wasn't aware of that, but also re:

> Using one commonly used configuration file gives us the advantage of having one single source of truth. It also means we don't have to remember every configuration file location for every tool we ever want to use.

I didn't see any existing configs in the file, this was in fact the first time I've encountered settings being put there, so I had to go look up what's up with that.

So it's not like my suggestion of using these djLint-specific files is me working against any established conventions for our projects or whatever, and I object to the insinuation it was outlandish. If this should be a rule, it'd help to officially make it one; there are other team members besides me who didn't use Poetry at all either until our recently-ish switch to it, and who might have looked into it even less than me since then.

> I don't get what the way a software is installed has to do with where it gets its configuration settings from.

It doesn't, I was expressly mentioning _transparency_, which is relevant to humans, not machines. If half the team doesn't know `pyproject.toml` is used to store settings for random tools that don't have any connection to Poetry, how would they know to go looking for them there if they wanted to change something about the workflow.

**Comment by koeaw** [comment]
Closing this as this repo is being phased out and all further discussion of linters for HTML, CSS and JS should probably happen over in [apis-core-rdf](https://github.com/acdh-oeaw/apis-core-rdf). The latter's files got formatted once, I believe, though I'm not sure if we're currently enforcing formatting on new/changed files included in PRs via GitHub actions.
